nutils/sample.py

```diff
@@ -0,0 +1,462 @@
+
+    Args
+    ____
+    arguments : :class:`dict`
+        Arguments for function evaluation.
+
+    Returns
+    -------
+    replaced : :class:`Integral`
+    '''
+
+    return Integral([di, function.replace_arguments(integrand, arguments)] for di, integrand in self._integrands.items())
+
+  def contains(self, name):
+    '''Test if target occurs in any of the integrands.
+
+    Args
+    ____
+    name : :class:`str`
+        Target name.
+
+    Returns
+    _______
+    iscontained : :class:`bool`
+    '''
+
+    try:
+      self._argshape(name)
+    except KeyError:
+      return False
+    else:
+      return True
+
+  def __add__(self, other):
+    if not isinstance(other, Integral):
+      return NotImplemented
+    assert self.shape == other.shape
+    integrands = self._integrands.copy()
+    for di, integrand in other._integrands.items():
+      try:
+        integrands[di] += integrand
+      except KeyError:
+        integrands[di] = integrand
+    return Integral(integrands.items())
+
+  def __neg__(self):
+    return Integral([di, -integrand] for di, integrand in self._integrands.items())
+
+  def __sub__(self, other):
+    return self + (-other)
+
+  def __mul__(self, other):
+    if not isinstance(other, numbers.Number):
+      return NotImplemented
+    return Integral([di, integrand * other] for di, integrand in self._integrands.items())
+
+  __rmul__ = __mul__
+
+  def __truediv__(self, other):
+    if not isinstance(other, numbers.Number):
+      return NotImplemented
+    return self.__mul__(1/other)
+
+  def _argshape(self, name):
+    assert isinstance(name, str)
+    shapes = {func.shape[:func.ndim-func._nderiv]
+      for func in function.Tuple(self._integrands.values()).dependencies
+        if isinstance(func, function.Argument) and func._name == name}
+    if not shapes:
+      raise KeyError(name)
+    assert len(shapes) == 1, 'inconsistent shapes for argument {!r}'.format(name)
+    shape, = shapes
+    return shape
+
+strictintegral = types.strict[Integral]
+
+@types.apply_annotations
+@cache.function
+def eval_integrals(*integrals: types.tuple[strictintegral], arguments:types.frozendict[types.strictstr,types.frozenarray]=None):
+  '''Evaluate integrals.
+
+  Evaluate one or several postponed integrals. By evaluating them
+  simultaneously, rather than using :func:`Integral.eval` on each integral
+  individually, integrations will be grouped per Sample and jointly executed,
+  potentially increasing efficiency.
+
+  Args
+  ____
+  integrals : :class:`tuple` of integrals
+      Integrals to be evaluated.
+  arguments : :class:`dict` (default: None)
+      Optional arguments for function evaluation.
+
+  Returns
+  -------
+  results : :class:`tuple` of arrays and/or :class:`nutils.matrix.Matrix` objects.
+  '''
+
+  gather = {}
+  for iint, integral in enumerate(integrals):
+    for di in integral._integrands:
+      gather.setdefault(di, []).append(iint)
+  retvals = [None] * len(integrals)
```

Comment:
Can we use `util.gather` here?

    gather = util.gather((di, iint) for iint, intergral in enumerate(integrals) for di in integral._integrands)